프로필 상세화면 구현

quessr opened the pull request:
- 프로필 상세 화면 ui 구현
- 10문10답 펼쳐서 보기/ 접기 기능 구현

quessr committed:
conflict 해결

#20

quessr committed:
Merge branch 'develop' into hsr/profile_detail_#20

# Conflicts:
#	app/build.gradle.kts

quessr committed:
커스텀 dotIndicator가 포함된 이미지 슬라이드 구현

#20
- 커스텀 ViewPagerAdapter를 통해 프로필 이미지를 위한 슬라이드 기능과 탭 간의 연동을 구현
- TabLayout에 커스텀 dotIndicator를 추가

quessr committed:
툴바의 가시성을 메인 프래그먼트에서만 표시하도록 조정

#20

quessr committed:
프로필디테일 레이아웃 구현 - 1차

#20
- 이름, 학교, 나이 섹션
- 배지, TEEN 카드 섹션
- 자기소개 섹션

quessr committed:
커스텀 QuestionAnswerView 구현

#20

quessr committed:
QuestionAnswerAdapter 구현 및 QuestionAnswerView를 RecyclerView로 변경

#20

quessr committed:
10문10답 RecyclerView 마지막 아이템에 '펼쳐서 보기' 옵션 추가 https://github.com/LayTheGroundWork/Eighteen_Android/issues/20

- 아이템 개수가 3개 이상일 때만 마지막 아이템에 '더 보기' 버튼이 포함된 뷰 사용
- 아이템 타입 분기 처리를 통해 일반 아이템과 마지막 아이템 레이아웃 구분

quessr committed:
glide 옵션 추가 https://github.com/LayTheGroundWork/Eighteen_Android/issues/20

quessr committed:
필요없는 코드 및 파일 삭제, 정리 https://github.com/LayTheGroundWork/Eighteen_Android/issues/20

quessr committed:
펼쳐서 보기를 누르면 모든 아이템을 표시하는 기능 추가  https://github.com/LayTheGroundWork/Eighteen_Android/issues/20

quessr committed:
Merge branch 'develop' into hsr/profile_detail_#20

# Conflicts:
#	app/build.gradle.kts
#	app/src/main/java/com/eighteen/eighteenandroid/presentation/profileDetail/ProfileDetailFragment.kt

kim0hoon commented on app/src/main/java/com/eighteen/eighteenandroid/presentation/MainActivity.kt:
```diff
@@ -40,6 +40,8 @@ class MainActivity : AppCompatActivity() {
             binding.bottomNavigationBar.visibility =
                 if (destination.id == R.id.fragmentMain) View.VISIBLE else View.GONE
+            binding.toolbar.root.visibility =
+                if (destination.id == R.id.fragmentMain) View.VISIBLE else View.GONE
```
bottomNavigationBar.isVisible = (true/false)로 사용하시면 간편해요
View.isVisible이라는 확장함수가 있습니다~

quessr replied:
> bottomNavigationBar.isVisible = (true/false)로 사용하시면 간편해요 View.isVisible이라는 확장함수가 있습니다~

수정했습니다 훨씬 깔끔하네요! 알려주셔서 감사합니다 🙏

kim0hoon commented on app/src/main/java/com/eighteen/eighteenandroid/presentation/profileDetail/ProfileDetailFragment.kt:
```diff
@@ -18,23 +21,65 @@ class ProfileDetailFragment :
 
     private val mediaDetailViewModel by viewModels<MediaDetailViewModel>()
-
+    private lateinit var viewPager: ViewPager2
+    private lateinit var tabLayout: TabLayout
```
뷰들은 binding을 통해서 접근 가능합니다!

quessr replied:
> 뷰들은 binding을 통해서 접근 가능합니다!

네 삭제했습니다! 🙏

kim0hoon commented on app/src/main/java/com/eighteen/eighteenandroid/presentation/profileDetail/QuestionAnswerAdapter.kt:
```diff
@@ -0,0 +1,87 @@
+import com.eighteen.eighteenandroid.databinding.QuestionAnswerItemBinding
+import com.eighteen.eighteenandroid.databinding.QuestionAnswerItemWithSeeMoreBinding
+
+class QuestionAnswerAdapter(private val items: List<Pair<String, String>>, private val recyclerView: RecyclerView) :
```
adapter에서 recyclerView를 받는건 위험할 수도 있습니다
adapter가 view보다 lifecycle이 길어서
외부에서 뷰 생명주기에 맞춰서(onViewCreated ~ onViewDestroy) adapter 참조를 풀어줘서 맞춰주거나 하셔야 해요

kim0hoon commented on app/src/main/java/com/eighteen/eighteenandroid/presentation/profileDetail/QuestionAnswerAdapter.kt:
```diff
@@ -0,0 +1,87 @@
+        private const val ITEM_COUNT_THRESHOLD = 2
+    }
+
+    private var showAllItems: Boolean = false
+    private fun showAllItems() {
+        showAllItems = true
+        notifyDataSetChanged()
+    }
```
보통 뷰에 표시할 내용들은 ViewModel에 저장해놓고 observe하는 방식으로 구현하는데
상태(펼쳐짐, 접힘)값도 ViewModel에 놓고 사용하시는게 좋아요
얻을 수 있는 이점이
1. 구성변경(configuration change)뒤에도 상태값이 유지가 되고
2. 네비게이션으로 다음 화면으로 넘어갔다가 돌아올 때도 상태를 유지할 수 있습니다!

그리고 diffUtil이나 ListAdapter도 한번 보시면 좋아요
RecyclerView 내용(펼치기/접기 상태도 포함)을 변경할 때 쉽게 update할 수 있어서 뷰쪽 작업에서 많이 사용하실거에요!

quessr replied:
@kim0hoon 
상세하게 알려주셔서 너무 감사합니다🙏

말씀해 주신대로, diffUtil과 ListAdapter내용을 알아보다가
궁금한 점이 있어서 질문 드립니다!
DiffUtil은 ListAdapter와 같이 사용할때가 더 효율적이라는 것은 알겠는데
ListAdapter에서 말고 RecyclerView.Adapter에서도 DiffCallback을 구현해서
DiffUtil을 사용할수가 있잖아요,,
그럼 RecyclerView.Adapter랑 DiffUtil을 사용하는 경우는 거의 없는걸까요?!?!

DiffUtil은 항상 ListAdapter랑만 같이 사용하시는지 궁금합니다!

kim0hoon replied:
ListAdapter는 편하게 쓰려고 RecyclerView Adapter를 상속받아 diffutil의 일부 기능들을 구현한 클래스에요
코드보시면 엄청 간단하게 되있는데
이걸 RecyclerView Adapter에 그대로 구현하시면 같은 기능을 합니다~

kim0hoon commented on app/src/main/res/layout/fragment_profile_detail.xml:
```diff
@@ -1,21 +1,242 @@
 <?xml version="1.0" encoding="utf-8"?>
-<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
+<ScrollView xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
```
이부분은 잘 기억이 안나는데 ScrollView 안에 RecyclerView를 두면 아마 스크롤이 꼬일거에요
몇가지 생각나는 방법으론
1. NestedScrollView로 변경하시거나
2. RecyclerView대신 ViewGroup으로 변경하고 bind시점에 ViewGroup을 clear하고 add하는 방법

두개가 있긴한데 1번은 여기는 작아서 상관없지만 RecyclerView 부분이 클 경우 지양해야 하고(모든 ViewHolder들을 생성해서 버벅입니다)

추가로 화면이 클 경우(스크롤이 길 경우) 화면 전체를 RecyclerView로 두고 각 요소들을 ViewHolder로 두는게 더 좋습니다
ViewHolder를 다른화면에서 재활용할 수 있고 한 파일 크기가 너무 커지지 않아서요

kim0hoon commented on app/src/main/res/layout/fragment_profile_detail.xml:
```diff
@@ -1,21 +1,242 @@
-        android:layout_gravity="top|end"/>
+        android:orientation="vertical">
+
+        <LinearLayout
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:orientation="vertical">
```
너무 깊지 않으면 상관없지만 뷰그룹 depth는 적게 두는게 좋습니다!
ConstraintLayout으로 대부분 구현 가능하긴 해요!

kim0hoon commented on app/src/main/java/com/eighteen/eighteenandroid/presentation/profileDetail/QuestionAnswerAdapter.kt:
```diff
@@ -0,0 +1,87 @@
+//        }
+        if (getItemViewType(position) == ITEM_TYPE_LAST) {
+            val item = items[position - 1]
+            (holder as LastViewHolder).bind(item.first, item.second)
+        } else {
+            val item = items[position]
+            (holder as NormalViewHolder).bind(item.first, item.second)
```
as보단 safecast(as?) 사용을 추천드립니다!

quessr committed:
visibility 설정을 View.isVisible 확장 함수를 사용하여 수정

https://github.com/LayTheGroundWork/Eighteen_Android/issues/20

quessr committed:
필요없는 코드 삭제

https://github.com/LayTheGroundWork/Eighteen_Android/issues/20

quessr committed:
10문10답 "펼쳐서 보기/접기" 기능 구현 #20

- 두 개의 ViewHolder를 사용하여 두 개의 레이아웃을 보여주던 방식을 하나의 ViewHolder로 통합
- 단일 ViewHolder 내에서 "펼쳐서 보기/접기" 토글 기능 구현

quessr committed:
폴더명 변경 #20

quessr committed:
Merge branch 'develop' into hsr/profile_detail_#20

# Conflicts:
#	app/src/main/res/values/colors.xml

quessr committed:
ProfileDetailViewModel 추가 #20

quessr committed:
프로필 상세화면 리팩토링 #20

- ViewModel과 LiveData를 사용하여 데이터 및 상태 관리

quessr committed:
Glide 이미지 로딩을 ImageLoader로 변경 #20

quessr committed:
View visibility 설정을 isVisible 속성으로 변경 #20

quessr committed:
ProfileDetailModel 추가 #20

quessr committed:
StateFlow와 setItems 메서드를 ProfileDetailModel을 사용하도록 수정 #20

- MutableStateFlow 타입을 List<Pair<String, String>>에서 List<ProfileDetailModel>로 변경
- setItems 메서드가 List<Pair<String, String>> 대신 List<ProfileDetailModel.Qna>를 받도록 수정

quessr committed:
setItems 호출 시 Pair 대신 ProfileDetailModel.Qna 사용하도록 수정 #20

- setItems 호출 시 List<Pair<String, String>> 대신 List<ProfileDetailModel.Qna> 사용하도록 수정

quessr committed:
ProfileDetailModel 타입 불일치 문제 해결 #20

quessr committed:
ProfileDetailModel에 따른 데이터 구조 및 초기화 방식 수정 #20

- ProfileDetailModel을 사용하여 프로필 상세 데이터 구조 정의
- initViewModel 함수 추가로 ViewModel 초기화 로직 분리

quessr committed:
아이템 바인딩 시 인덱스 초과 문제 해결 #20

quessr committed:
Merge branch 'develop' into hsr/profile_detail_#20

# Conflicts:
#	app/src/main/java/com/eighteen/eighteenandroid/presentation/MainActivity.kt

quessr committed:
코드 정리 #20

- 불필요한 import 및 사용되지 않는 코드 삭제

quessr committed:
RecyclerView.Adapter에서 ListAdapter로 변경하여 DiffUtil 적용 #20

- ProfileDetailModel.Qna 클래스에 id 필드 추가

quessr committed:
Merge branch 'develop' into hsr/profile_detail_#20

quessr committed:
QnA 리스트에 ID 추가 #20

quessr committed:
QnA 뷰 바인딩의 레이아웃 파일 이름을 ItemProfileDetailQnaBinding으로 변경 #20

quessr committed:
프로필 상세 화면 레이아웃 ScrollView에서 RecyclerView로 변경 #20

quessr committed:
profile detail info 레이아웃 추가 #20

- 프로필 이름 및 학교, 나이 정보 레이아웃 추가

quessr committed:
profile detail images 레이아웃 추가 #20

- 뷰페이저와 탭 레이아웃을 사용한 이미지 선택 레이아웃 추가

quessr committed:
ProfileDetailModel 클래스에 id 필드를 추가 #20

quessr committed:
ProfileDetailViewHolder 클래스 추가 #20

quessr committed:
ProfileDetailAdapter 구현 #20

quessr committed:
ProfileDetailFragment에 ProfileDetailAdapter 적용 및 초기 데이터 설정 로직 추가 #20

quessr committed:
레이아웃 수정 #20

- 중첩된 ConstraintLayout을 제거

quessr committed:
profile detail badge and teen 레이아웃 추가 #20

quessr committed:
profile detail introduction 레이아웃 추가 #20

quessr committed:
profile detail qna 레이아웃 추가 #20

quessr committed:
qna item 레이아웃 추가 #20

quessr committed:
BadgeAndTeen 및 QnaList 데이터 클래스 추가 #20

quessr committed:
BadgeAndTeen, Introduction, Qna ViewHolder 타입 추가 #20

quessr committed:
BADGE_AND_TEEN,  INTRODUCTION, QNA_LIST 뷰 타입 추가 #20

quessr committed:
Qna 처리에서 QnaList 처리로 변경 #20

quessr committed:
코드정리 및 QnaList 데이터 추가 #20

quessr committed:
item qna 레이아웃 수정 및  itme see more 레이아웃 추가 #20

- item qna 레이아웃에서 itme see more 레이아웃 분리

quessr committed:
Merge branch 'develop' into hsr/profile_detail_#20

quessr committed:
QnaListTitle 및 QnaToggle 클래스 추가 #20

quessr committed:
qna title 레이아웃 추가 #20

quessr committed:
스타일 수정 #20

quessr committed:
Qna 관련 ViewHolder 추가 및 수정 #20

- QnaTitle, QnaToggle 추가

quessr committed:
Qna 항목 관련 ViewModel 로직 개선 및 QnaToggle, QnaListTitle 처리 추가 #20

quessr committed:
Qna 관련 데이터 처리 로직 업데이트 #20

quessr committed:
레이아웃 및 스타일 수정 #20

- 불필요한 주석 제거

quessr committed:
drawable 추가 #20

quessr committed:
배지, TEEN background image 추가 및 레이아웃 수정 #20

quessr committed:
파일 이름 수정, 좋아요 추가 및 레이아웃 수정 #20

quessr committed:
코드 정리 #20

quessr committed:
스타일 수정 #20

quessr committed:
Qna 항목의 배경 및 하단 마진 적용 #20

quessr committed:
RecyclerView의 itemAnimator 제거하여 애니메이션 효과 제거 #20

quessr committed:
필요없는 파일 삭제 #20

quessr committed:
Merge branch 'develop' into hsr/profile_detail_#20

quessr committed:
values/colors.xml에 정의된 색상 적용 #20

quessr committed:
QnA 항목의 답변 텍스트가 질문 텍스트로 잘못 설정된 문제 수정 #20

- QnA 초기 데이터 뱐경

quessr committed:
values/typo.xml 적용 #20

quessr committed:
필요없는 파일 및 주석 삭제 #20

quessr committed:
profile detail strings 추가

quessr committed:
values/strings.xml 적용

quessr commented:
> * 프로필 상세 화면 ui 구현
> * 10문10답 펼쳐서 보기/ 접기 기능 구현

@kim0hoon 
리뷰 주신 내용 전부 반영 했습니다!
머지 해도 될지 확인 부탁드리겠습니다 🙏

kim0hoon commented on app/src/main/java/com/eighteen/eighteenandroid/presentation/profiledetail/ProfileDetailAdapter.kt:
```diff
@@ -0,0 +1,161 @@
+        val context = parent.context
+        val inflater = LayoutInflater.from(context)
+
+        fun <VB : ViewBinding> inflaterBinding(bindingInflater: (LayoutInflater, ViewGroup, Boolean) -> VB) =
```
inflate 오타나신 것 같아요
함수는 동사로 부탁드립니다~

quessr replied:
네 수정했습니다~

kim0hoon commented on app/src/main/java/com/eighteen/eighteenandroid/presentation/profiledetail/ProfileDetailAdapter.kt:
```diff
@@ -0,0 +1,161 @@
+//            ITEM_TYPE_QNA_LIST -> {
+//                val binding = inflaterBinding(ItemProfileDetailQnaBinding::inflate)
+//                ProfileDetailViewHolder.Qna(binding, lifecycleOwner, viewModel)
+//            }
```
안쓰는 부분은 지워주세요~

quessr replied:
네 삭제했습니다~

kim0hoon commented on app/src/main/java/com/eighteen/eighteenandroid/presentation/profiledetail/ProfileDetailFragment.kt:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    private fun initViewModel() {
+        lifecycleScope.launch {
+            profileDetailViewModel.items.collectLatest { details ->
+                adapter.submitList(details)
+            }
+        }
```
Fragment에서 뷰가 포함된 작업은 viewLifecycleOwner를 따르셔야 해요
Fragment 생명주기(onCreate~onDestory)랑 view의 생명주기(onCreateView ~ onDestroyView)랑 달라서 메모리릭이 발생합니다!

추가로 collect 작업은 repeatOnLifecycle(State.START) (onStart~onPause 범위) 안에서 실행해주시는게 좋아요
홈버튼 눌러서 밖으로 나갈 때 collect가 발생해서 백그라운드에서 불필요한 리소스 낭비가 발생해서요
https://reco-dy.tistory.com/12 여기 참고해보시면 좋을 것 같아요

quessr replied:
> Fragment에서 뷰가 포함된 작업은 viewLifecycleOwner를 따르셔야 해요 Fragment 생명주기(onCreate~onDestory)랑 view의 생명주기(onCreateView ~ onDestroyView)랑 달라서 메모리릭이 발생합니다!
> 
> 추가로 collect 작업은 repeatOnLifecycle(State.START) (onStart~onPause 범위) 안에서 실행해주시는게 좋아요 홈버튼 눌러서 밖으로 나갈 때 collect가 발생해서 백그라운드에서 불필요한 리소스 낭비가 발생해서요 https://reco-dy.tistory.com/12 여기 참고해보시면 좋을 것 같아요

우와.. 전혀 몰랐네요 알려주셔서 감사합니다 🙏
혹시 모르니까 맞게 수정한건지 확인 부탁드립니다..!

kim0hoon replied:
수고하셨습니다~ 👍

kim0hoon commented on app/src/main/java/com/eighteen/eighteenandroid/presentation/profiledetail/ProfileDetailViewModel.kt:
```diff
@@ -0,0 +1,95 @@
+    private val _items = MutableStateFlow<List<ProfileDetailModel>>(emptyList())
+    val items: StateFlow<List<ProfileDetailModel>> get() = _items
+
+    private val _showAllItems = MutableStateFlow(false)
+    val showAllItems: StateFlow<Boolean> get() = _showAllItems
+
+    private val _qnaItems = MutableStateFlow<List<ProfileDetailModel.Qna>>(emptyList())
+    val qnaItems: StateFlow<List<ProfileDetailModel.Qna>> get() = _qnaItems
```
하나의 모델로 합치신 것 같은데 이부분도 안쓰는게 맞을까요??

kim0hoon commented on app/src/main/java/com/eighteen/eighteenandroid/presentation/profiledetail/ProfileDetailAdapter.kt:
```diff
@@ -0,0 +1,161 @@
+import com.eighteen.eighteenandroid.presentation.profiledetail.viewholder.ProfileDetailViewHolder
+
+class ProfileDetailAdapter(
+    private val lifecycleOwner: LifecycleOwner,
+    private val viewModel: ProfileDetailViewModel
```
혹시 여기는 사용하시는 부분인가요??

quessr replied:
삭제 했습니다~

kim0hoon reviewed (commented):
StateFlow는 생소하실텐데 고생 많으셨습니다 ㅎㅎㅎ
크게 고칠만한 부분은 이제 없는 것 같네요

quessr commented:
> StateFlow는 생소하실텐데 고생 많으셨습니다 ㅎㅎㅎ 크게 고칠만한 부분은 이제 없는 것 같네요

와.. 엄청 세심하게 봐주셨네요 감사합니다 ㅎㅎㅎ!

quessr committed:
변수명 수정 #20

quessr committed:
주석 및 로그 삭제 #20

quessr committed:
필요없는 코드 삭제 #20

kim0hoon commented on app/src/main/java/com/eighteen/eighteenandroid/presentation/profiledetail/viewholder/ProfileDetailViewHolder.kt:
```diff
@@ -0,0 +1,114 @@
+
+    class QnaToggle(
+        val binding: ItemSeeMoreBinding,
+        private val profileDetailViewModel: ProfileDetailViewModel
```
파라미터로 뷰모델을 넘기시기보단 toggle 함수만 넘기시는게 어떨까요?
뷰홀더가 뷰모델 전체를 알고있을 필요가 없어서요!

quessr replied:
> 파라미터로 뷰모델을 넘기시기보단 toggle 함수만 넘기시는게 어떨까요? 뷰홀더가 뷰모델 전체를 알고있을 필요가 없어서요!

오.. 그러네요?! 그래야 뷰홀더가 더 명확해지네요,, 감사합니다 최고최고 👍
말씀 해 주신대로 수정했습니다!

quessr committed:
lifecycleScope를 viewLifecycleOwner.lifecycleScope로 변경 및 repeatOnLifecycle 사용 #20

quessr committed:
QnaToggle 뷰홀더에 뷰모델 대신 toggle 함수를 전달하도록 수정 #20

quessr committed:
gradient 추가 및 뷰페이저에 상단 및 하단에 적용  #20
